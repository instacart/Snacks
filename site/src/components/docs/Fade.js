import React, {Fragment} from 'react'
import Playground from '../Playground'

export default () => (

Example:

<Playground>
{`
    initialState = { isOpen: false };
    width = 250;

    <div>
      <Button onClick={() => setState({ isOpen: !state.isOpen })} style={{ marginBottom: 10 }}>Toggle</Button>
      <div style={{ overflow: 'hidden'}}>
        <Fade in={state.isOpen} width={width + 42}>
          <div style={{padding: 20, display: 'flex', alignItems:'center', background: '#43B02A', flexDirection: 'column', width: width, border: '1px solid black' }}>
            <LoadingBox shape='circle' />
            <LoadingBox shape='line' />
            <LoadingBox shape='line' />
            <LoadingBox shape='line' />
          </div>
        </Fade>
      </div>
      <div style={{ marginTop: 10 }}>
        Content below
      </div>
    </div>
`}
</Playground>

)
