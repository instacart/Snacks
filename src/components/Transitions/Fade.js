import React, { PureComponent } from 'react'
import Transition from 'react-transition-group/Transition'
import PropTypes from 'prop-types'

const TIMEOUT = 0
const TRASITION_TIME = 400
const TIMING_FUNCTION = 'ease-in-out'
const START_OPACITY = 0
const END_OPACITY = 1

const OPACITY_DEFAULT = {
  start: 0,
  end: 1,
}

class Fade extends PureComponent {
  static propTypes = {
    /**
     * A convenience prop that enables or disabled appear animations for
     * all children. Note that specifying this will override any defaults
     * set on individual children Transitions.
     */
    appear: PropTypes.bool,

    children: PropTypes.node,

    className: PropTypes.string,

    /** Show the component; triggers the enter or exit states */
    in: PropTypes.bool,

    /**
     * Settings for opacity during animation.
     *
     * Default start: 0
     * Default end: 1500 (If content's height is larger than 1500, pass the content's height here )
     */
    opacity: PropTypes.shape({
      start: PropTypes.number,
      end: PropTypes.number,
    }),

    /**
     * Optional style overrides merged into emotion css
     *
     * @deprecated
     * This prop exists for backwards compatibility and will be
     * removed in a future version
     */
    style: PropTypes.object,

    /** Delay in milliseconds until animation start. */
    timeout: PropTypes.number,

    /** Time of animation in milliseconds. */
    transitionTime: PropTypes.number,

    /** Name of the transition-timing-function CSS property. */
    timingFunction: PropTypes.oneOf(['ease', 'ease-in', 'ease-out', 'ease-in-out', 'linear']),
  }

  static defaultProps = {
    timeout: TIMEOUT,
    transitionTime: TRASITION_TIME,
    timingFunction: TIMING_FUNCTION,
    opacity: OPACITY_DEFAULT,
    style: {},
    in: true,
    appear: true,
  }

  get opacity() {
    const { start = START_OPACITY, end = END_OPACITY } = this.props.opacity
    return {
      start,
      end,
    }
  }

  get transitionStyles() {
    const { start: opacityStart, end: opacityEnd } = this.opacity

    return {
      entering: {
        opacity: opacityStart,
      },
      entered: {
        opacity: opacityEnd,
      },
    }
  }

  get initialStyles() {
    const { transitionTime, timingFunction } = this.props
    const { start: opacityStart } = this.opacity
    return {
      opacity: opacityStart,
      transition: `all ${transitionTime}ms ${timingFunction}`,
    }
  }

  renderChild = state => {
    const { className, style, children } = this.props
    const styles = {
      ...style,
      ...this.initialStyles,
      ...this.transitionStyles[state],
    }
    return (
      <div className={className} css={styles}>
        {children}
      </div>
    )
  }

  render() {
    const { in: inProp, timeout, appear } = this.props
    return (
      <Transition in={inProp} appear={appear} timeout={timeout}>
        {this.renderChild}
      </Transition>
    )
  }
}

export default Fade
