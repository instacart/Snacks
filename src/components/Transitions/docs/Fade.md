Example:

```jsx static
const initialState = { isOpen: false }
const width = 250

<div>
  <Button onClick={() => setState({ isOpen: !state.isOpen })} style={{ marginBottom: 10 }}>Toggle</Button>
  <div style={{ overflow: 'hidden'}}>
    <Fade in={state.isOpen} width={width + 42}>
      <div style={{padding: 20, display: 'flex', alignItems:'center', background: '#43B02A', flexDirection: 'column', width: width, border: '1px solid black' }}>
        <LoadingBox shape="circle" />
        <LoadingBox shape="line" />
        <LoadingBox shape="line" />
        <LoadingBox shape="line" />
      </div>
    </Fade>
  </div>
  <div style={{ marginTop: 10 }}>
    Content below
  </div>
</div>
```
