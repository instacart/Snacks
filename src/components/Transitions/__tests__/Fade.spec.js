import React from 'react'
import renderer from 'react-test-renderer'
import Fade from '../Fade'

jest.mock('react-dom', () => ({
  findDOMNode: () => ({
    getContext: jest.fn(),
  }),
}))

it('renders Fade in correctly', () => {
  const tree = renderer
    .create(
      <Fade in>
        <h3> Content </h3>
      </Fade>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders Fade correctly with custom props', () => {
  const opacity = {
    start: 0.2,
    end: 0.8,
  }
  const tree = renderer
    .create(
      <Fade in axis="y" style={{ background: 'red' }} transitionTime={500} opacity={opacity}>
        <h3> Content </h3>
      </Fade>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
