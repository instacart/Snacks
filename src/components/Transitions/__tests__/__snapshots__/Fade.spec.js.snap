// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Fade correctly with custom props 1`] = `
<div
  data-radium={true}
>
  <div
    style={
      Object {
        "background": "red",
        "opacity": 0.2,
        "transition": "all 500ms ease-in-out",
      }
    }
  >
    <h3>
       Content 
    </h3>
  </div>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;

exports[`renders Fade in correctly 1`] = `
<div
  data-radium={true}
>
  <div
    style={
      Object {
        "opacity": 0,
        "transition": "all 400ms ease-in-out",
      }
    }
  >
    <h3>
       Content 
    </h3>
  </div>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;
